Expose callback that allows focus traversal customization

JellyO1 opened the pull request:
This PR exposes a requestFocusCallback on `FocusTraversalPolicy` and it's inheritors.

Fixes #83175.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

JellyO1 committed:
Expose callback that allows focus traversal customization

JellyO1 committed:
Remove unnecessary cast

JellyO1 committed:
Replace _focusAndEnsureVisible calls with requestFocusCallback

gspencergoog commented on packages/flutter/lib/src/widgets/focus_traversal.dart:
```diff
@@ -38,11 +38,24 @@ BuildContext? _getAncestor(BuildContext context, {int count = 1}) {
 
+/// Signature for the callback that's called when a traversal policy is
+/// requesting focus.
+typedef RequestFocusCallback = void Function(
```
This should probably have a name that reflects that it is part of focus traversal, and not some version of the requestFocus on a focus node. Maybe `TraversalRequestFocusCallback`?

gspencergoog commented on packages/flutter/lib/src/widgets/focus_traversal.dart:
```diff
@@ -150,7 +163,13 @@ enum TraversalEdgeBehavior {
   /// Abstract const constructor. This constructor enables subclasses to provide
   /// const constructors so that they can be used in const expressions.
-  const FocusTraversalPolicy();
+  const FocusTraversalPolicy({
```
You'll need to describe the argument and what it does in the constructor API documentation.

gspencergoog commented on packages/flutter/lib/src/widgets/focus_traversal.dart:
```diff
@@ -150,7 +163,13 @@ enum TraversalEdgeBehavior {
+  }) : requestFocusCallback = requestFocusCallback ?? _focusAndEnsureVisible;
+
+  /// The callback used to move the focus from one focus node to another when
+  /// traversing them using a keyboard.
```
This needs to talk about what the default behavior is, and when it is invoked.

gspencergoog commented on packages/flutter/lib/src/widgets/focus_traversal.dart:
```diff
@@ -38,11 +38,24 @@ BuildContext? _getAncestor(BuildContext context, {int count = 1}) {
 void _focusAndEnsureVisible(
```
If we're going to make a public value where this is the default, perhaps it makes sense to make this a static public function on FocusTraversalPolicy, called `defaultTraversalRequestFocusCallback`? Then people could implement their callback to call the default if all they want to do is log analytics or something.

JellyO1 replied:
Perhaps, but I'm partial to keeping this function as it is and duplicating it as a static one on FocusTraversalPolicy since the method is also used in `RequestFocusAction`.

gspencergoog commented on packages/flutter/lib/src/widgets/focus_traversal.dart:
```diff
@@ -917,6 +936,9 @@ mixin DirectionalFocusTraversalPolicyMixin on FocusTraversalPolicy {
 ///    explicitly using [FocusTraversalOrder] widgets.
 class WidgetOrderTraversalPolicy extends FocusTraversalPolicy with DirectionalFocusTraversalPolicyMixin {
+  /// Constructs a traversal policy that orders widgets for keyboard traversal
+  /// based on the widget hierarchy order.
```
You'll need to describe the argument here too.  You might want to make a template with `{@template ...}` and `{@macro ...}` (see examples elsewhere in the framework).

gspencergoog commented on packages/flutter/lib/src/widgets/focus_traversal.dart:
```diff
@@ -1084,6 +1106,9 @@ class _ReadingOrderDirectionalGroupData with Diagnosticable {
 ///  * [OrderedTraversalPolicy], a policy that describes the order
 ///    explicitly using [FocusTraversalOrder] widgets.
 class ReadingOrderTraversalPolicy extends FocusTraversalPolicy with DirectionalFocusTraversalPolicyMixin {
+  /// Constructs a traversal policy that orders the widgets in "reading order".
```
Ditto here about documenting the arguments.

gspencergoog reviewed (commented):
This is a nice idea, thanks for the PR! When I wrote It, I thought about doing something similar, but decided to wait until there was a clear use case to make it public. Do you have a specific use case in mind?

**This PR will need to have tests before we can merge it.**

JellyO1 commented:
Thanks for the feedback, I'll make the changes this weekend. My current use case is that I'm developing an Android TV application with stripe navigation and in the design requirements they require that the selected element always stays at the start so that you can clearly see what comes ahead.
I'll also write the tests which will probably require further feedback as I'm not sure what exactly how and what I'm going to test.

JellyO1 committed:
Create static traversal request callback method

JellyO1 committed:
Document the requestFocusCallback argument

JellyO1 committed:
Change TraversalRequestFocusCallback named parameters to nullable

JellyO1 committed:
Create tests to verify that the callback is called on all instances

goderbauer commented:
(triage) @JellyO1 Do you still have plans to follow up on the feedback given above?

JellyO1 committed:
Remove duplicated code in favor of using FocusTraversalPolicy static implementation

JellyO1 committed:
Improve constructor and callback documentation

gspencergoog commented on packages/flutter/lib/src/widgets/focus_traversal.dart:
```diff
@@ -1712,7 +1754,7 @@ class RequestFocusIntent extends Intent {
 class RequestFocusAction extends Action<RequestFocusIntent> {
   @override
   void invoke(RequestFocusIntent intent) {
```
The `RequestFocusIntent` could take an optional `TraversalRequestFocusCallback` in its constructor, which you could call here.  That would let the `RequestFocusAction` be customizable in the same way.

Of course, It would also need a test.  Feel free to leave that for another PR if you don't have time.

JellyO1 replied:
That's a good catch, I'll do it on this PR since it makes the most sense.

gspencergoog reviewed (approved):
![32384589-a60f0e74-c078-11e7-9bc1-e5b5287aea9d](https://user-images.githubusercontent.com/8867023/37737237-66e9c9dc-2d10-11e8-8797-13a405eace49.png)

JellyO1 committed:
Change RequestFocusIntent to take optional TraversalRequestFocusCallback

This commit creates an optional TraversalRequestFocusCallback parameter
on `RequestFocusIntent` while keeping the original behaviour intact.
This allows you to change the behaviour of the focus request in the same
way you can change it on the `FocusTraversalPolicy`.

gspencergoog reviewed (approved):
![32384589-a60f0e74-c078-11e7-9bc1-e5b5287aea9d](https://user-images.githubusercontent.com/8867023/37737237-66e9c9dc-2d10-11e8-8797-13a405eace49.png) (again)

loic-sharma commented on packages/flutter/lib/src/widgets/focus_traversal.dart:
```diff
@@ -35,13 +35,15 @@ BuildContext? _getAncestor(BuildContext context, {int count = 1}) {
-  node.requestFocus();
-  Scrollable.ensureVisible(node.context!, alignment: 1.0, alignmentPolicy: alignmentPolicy);
-}
+/// Signature for the callback that's called when a traversal policy is
+/// requesting focus.
```
Consider: 
```suggestion
/// Signature for the callback that's called when a traversal policy
/// requests focus.
```

loic-sharma commented on packages/flutter/lib/src/widgets/focus_traversal.dart:
```diff
@@ -150,7 +152,37 @@ enum TraversalEdgeBehavior {
-  const FocusTraversalPolicy();
+  ///
+  /// {@template flutter.widgets.FocusTraversalPolicy.requestFocusCallback}
+  /// The `requestFocusCallback` can be used to override the default behaviour
```
Please prefer US english. See: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#spell-words-in-identifiers-and-comments-correctly

```suggestion
  /// The `requestFocusCallback` can be used to override the default behavior
```

JellyO1 replied:
Thanks for the heads-up, English is not my first language and honestly sometimes the nuances between UK and US English go over my head.

loic-sharma commented on packages/flutter/lib/src/widgets/focus_traversal.dart:
```diff
@@ -150,7 +152,37 @@ enum TraversalEdgeBehavior {
+        Curve? curve,
+      }) {
+    node.requestFocus();
+    Scrollable.ensureVisible(node.context!, alignment: alignment ?? 1.0,
+        alignmentPolicy: alignmentPolicy ?? ScrollPositionAlignmentPolicy.explicit,
+        duration: duration ?? Duration.zero,
+        curve: curve ?? Curves.ease);
```
```suggestion
    Scrollable.ensureVisible(
      node.context!, alignment: alignment ?? 1.0,
      alignmentPolicy: alignmentPolicy ?? ScrollPositionAlignmentPolicy.explicit,
      duration: duration ?? Duration.zero,
      curve: curve ?? Curves.ease,
    );
```

loic-sharma commented on packages/flutter/lib/src/widgets/focus_traversal.dart:
```diff
@@ -150,7 +152,37 @@ enum TraversalEdgeBehavior {
+  /// The default value for [requestFocusCallback].
+  /// Requests focus from `node` and ensures the node is visible
+  /// by calling [Scrollable.ensureVisible].
+  static void defaultTraversalRequestFocusCallback(
+      FocusNode node, {
+        ScrollPositionAlignmentPolicy? alignmentPolicy,
+        double? alignment,
+        Duration? duration,
+        Curve? curve,
+      }) {
```
```suggestion
  static void defaultTraversalRequestFocusCallback(
    FocusNode node, {
    ScrollPositionAlignmentPolicy? alignmentPolicy,
    double? alignment,
    Duration? duration,
    Curve? curve,
  }) {
```

Piinks committed:
Update packages/flutter/lib/src/widgets/focus_traversal.dart

Co-authored-by: Loïc Sharma <737941+loic-sharma@users.noreply.github.com>

Piinks committed:
Update packages/flutter/lib/src/widgets/focus_traversal.dart

Co-authored-by: Loïc Sharma <737941+loic-sharma@users.noreply.github.com>

Piinks committed:
Update packages/flutter/lib/src/widgets/focus_traversal.dart

Co-authored-by: Loïc Sharma <737941+loic-sharma@users.noreply.github.com>

Piinks committed:
Update packages/flutter/lib/src/widgets/focus_traversal.dart

Co-authored-by: Loïc Sharma <737941+loic-sharma@users.noreply.github.com>

Piinks reviewed (approved):
Welcome @JellyO1, thanks for contributing! 

![Flutter_LGTM](https://user-images.githubusercontent.com/16964204/169392212-02bb74fe-c6b9-4eb7-aa62-6ddb43a70d5b.png)

gspencergoog commented on packages/flutter/lib/src/widgets/focus_traversal.dart:
```diff
@@ -150,7 +152,39 @@ enum TraversalEdgeBehavior {
+  /// {@template flutter.widgets.FocusTraversalPolicy.requestFocusCallback}
+  /// The `requestFocusCallback` can be used to override the default behavior
+  /// of the focus requests. If `requestFocusCallback`
+  /// is null, it defaults to [defaultTraversalRequestFocusCallback].
```
```suggestion
  /// is null, it defaults to [FocusTraversalPolicy.defaultTraversalRequestFocusCallback].
```

JellyO1 committed:
Update packages/flutter/lib/src/widgets/focus_traversal.dart

Co-authored-by: Greg Spencer <gspencergoog@users.noreply.github.com>

goderbauer commented:
(Triage) Looks like some tests are failing. @JellyO1 can you take a look?

JellyO1 commented:
Yes, I'll take care of it.

JellyO1 committed:
Fix broken focus_traversal requestFocusCallback tests
